src/main/scala/cromwell/binding/WdlExpression.scala

```diff
@@ -1,41 +1,43 @@
 package cromwell.binding
 
-import java.util
-
-import cromwell.binding.types.{WdlFloatType, WdlStringType, WdlIntegerType, WdlObjectType}
-import cromwell.binding.values.{WdlString, WdlFloat, WdlInteger, WdlValue}
+import cromwell.binding.values._
 import cromwell.parser.WdlParser
-import cromwell.parser.WdlParser.{AstList, Ast, AstNode, Terminal}
+import cromwell.parser.WdlParser.{Ast, AstList, AstNode, Terminal}
+
 import scala.collection.JavaConverters._
-import scala.util.{Success, Failure, Try}
+import scala.util.{Success, Try}
 
 class WdlExpressionException(message: String = null, cause: Throwable = null) extends RuntimeException(message, cause)
 
 object WdlExpression {
   val parser = new WdlParser()
 
+  /** Maps from a locally qualified name to a WdlValue. */
+  // TODO change this to String => Option[WdlValue]
+  type LookupFunction = String => WdlValue
+
   def binaryOperators = Set(
     "Add", "Subtract", "Multiply", "Divide", "Remainder",
     "GreaterThan", "LessThan", "GreaterThanOrEqual", "LessThanOrEqual",
     "Equals", "NotEquals", "LogicalAnd", "LogicalOr"
   )
 
-  def evaluate(ast: AstNode, lookup: String => WdlValue, functions: WdlFunctions): WdlValue = {
+  def evaluate(ast: AstNode, lookup: LookupFunction, functions: WdlFunctions): Try[WdlValue] = {
     ast match {
-      case t: Terminal if t.getTerminalStr == "identifier" => lookup(t.getSourceString)
-      case t: Terminal if t.getTerminalStr == "integer" => WdlInteger(t.getSourceString.toInt)
-      case t: Terminal if t.getTerminalStr == "float" => WdlFloat(t.getSourceString.toFloat)
-      case t: Terminal if t.getTerminalStr == "string" => WdlString(t.getSourceString)
+      case t: Terminal if t.getTerminalStr == "identifier" => Success(lookup(t.getSourceString))
```

Comment:
I think it'll actually become `String => Option[WdlValue]` since this is essentially an engine-unaware lookup into the symbol store, right?